runtime/src/main/java/org/corfudb/runtime/view/LayoutView.java

```diff
@@ -126,66 +124,64 @@ public Layout prepare(long epoch, long rank)
                     return cf;
                 })
                 .toArray(CompletableFuture[]::new);
-        LayoutPrepareResponse[] acceptList;
+
+        long promises = 0L;
         long timeouts = 0L;
+        long outranks = 0L;
         long wrongEpochRejected = 0L;
-        while (true) {
-            // do we still have enough for a quorum?
-            if (prepareList.length < getQuorumNumber()) {
-                log.debug("prepare: Quorum unreachable, remaining={}, required={}", prepareList,
-                        getQuorumNumber());
-                throw new QuorumUnreachableException(prepareList.length, getQuorumNumber());
+        long outrankedRecord = Long.MIN_VALUE;
+        for (CompletableFuture cf : prepareList) {
+            if (promises >= getQuorumNumber()) {
+                break;
             }
 
-            // wait for someone to complete.
             try {
-                CFUtils.getUninterruptibly(CompletableFuture.anyOf(prepareList),
-                        OutrankedException.class, TimeoutException.class, NetworkException.class,
-                        WrongEpochException.class);
+                CFUtils.getUninterruptibly(cf, OutrankedException.class, TimeoutException.class,
```

Comment:
Great catch, this is so subtle. Using `CompletableFuture.anyOf` will return a `CompletableFuture` with a `CompletionException` on failure, but `getUninterruptibly` only operates on `Future` and thus only checks `ExecutionException`

Reply:
Why are we doing this ?
If the CFUtils is stuck on the wrong response it will wait even when the quorum is complete ?

Reply:
> Great catch, this is so subtle. Using `CompletableFuture.anyOf` will return a `CompletableFuture` with a `CompletionException` on failure, but `getUninterruptibly` only operates on `Future` and thus only checks `ExecutionException`

In CompletableFuture.java, get will call this function.
`    /**
     * Reports result using Future.get conventions.
     */
    private static <T> T reportGet(Object r)
        throws InterruptedException, ExecutionException {
        if (r == null) // by convention below, null means interrupted
            throw new InterruptedException();
        if (r instanceof AltResult) {
            Throwable x, cause;
            if ((x = ((AltResult)r).ex) == null)
                return null;
            if (x instanceof CancellationException)
                throw (CancellationException)x;
            if ((x instanceof CompletionException) &&
                (cause = x.getCause()) != null)
                x = cause;
            throw new ExecutionException(x);
        }
        @SuppressWarnings("unchecked") T t = (T) r;
        return t;
    }`

So getUninterruptibly is fine here.
I also did an experiment that blocks all prepare requests and timeout counter was bumped once.